File: java/observer_pattern/ObserverPattern.java
import java.util.ArrayList;
import java.util.List;

interface ObserverInterface {
    void update(String data);
}

interface ObservableInterface {
    void add(ObserverInterface obs);

    void notify(String data);
}

class RelevanceObservable implements ObservableInterface {
    private final List<ObserverInterface> observers = new ArrayList<>();

    @Override
    public void add(ObserverInterface obs) {
        observers.add(obs);
    }

    @Override
    public void notify(String data) {
        for (ObserverInterface observer : observers) {
            observer.update(data);
        }
    }
}

class RedisObserver implements ObserverInterface {
    private String data;

    @Override
    public void update(String data) {
        this.data = data;
        System.out.println("Redis data updated: " + data);
    }
}

class MongoObserver implements ObserverInterface {
    private String data;

    @Override
    public void update(String data) {
        this.data = data;
        System.out.println("Mongo data updated: " + data);
    }
}

public class ObserverPattern {
    public static void main(String[] args) {
        RelevanceObservable relevanceObservable = new RelevanceObservable();

        RedisObserver redisClient = new RedisObserver();
        MongoObserver mongoClient = new MongoObserver();

        relevanceObservable.add(redisClient);
        relevanceObservable.add(mongoClient);

        relevanceObservable.notify("data 1");
        relevanceObservable.notify("data 2");
    }
}

File: go/observer_pattern.go
package main

import "fmt"

type ObservableInterface interface {
	Add(obs ObserverInterface)
	Notify(data string)
}

type RelevanceObservable struct {
	observers []ObserverInterface
}

func (ro *RelevanceObservable) Add(obs ObserverInterface) {
	ro.observers = append(ro.observers, obs)
}

func (ro *RelevanceObservable) Notify(data string) {
	for _, observer := range ro.observers {
		observer.Update(data)
	}
}

type ObserverInterface interface {
	Update(data string)
}

type RedisObserver struct {
	data string
}

func (o *RedisObserver) Update(data string) {
	o.data = data
	fmt.Printf("Redis data updated: %s\n", data)
}

type MongoObserver struct {
	data string
}

func (o *MongoObserver) Update(data string) {
	o.data = data
	fmt.Printf("Mongo data updated: %s\n", data)
}

func main() {
	relevanceObserable := RelevanceObservable{}
	
	redisClient := RedisObserver{}
	mongoClient := MongoObserver{}

	relevanceObserable.Add(&redisClient)
	relevanceObserable.Add(&mongoClient)

	relevanceObserable.Notify("data 1")
	relevanceObserable.Notify("data 2")
}
